biomaj_download/download/interface.py

```diff
@@ -231,17 +254,80 @@ def download_or_copy(self, available_files, root_dir, check_exists=True):
+        '''
+        raise NotImplementedError()
+
+    def download(self, local_dir, keep_dirs=True):
         '''
         Download remote files to local_dir
 
         :param local_dir: Directory where files should be downloaded
         :type local_dir: str
+        :param keep_dirs: keep file name directory structure or copy file in local_dir directly
+        :param keep_dirs: bool
         :return: list of downloaded files
         '''
-        pass
+        self.logger.debug(self.__class__.__name__ + ':Download')
+        nb_files = len(self.files_to_download)
+        cur_files = 1
+        # Change working directory (needed to run commands)
+        try:
+            os.chdir(self.offline_dir)
+        except TypeError:
+            self.logger.error(self.__class__.__name__ + ":Download:Could not find offline_dir")
+        # Should we skip test of archives
+        skip_check_uncompress = os.environ.get('UNCOMPRESS_SKIP_CHECK', None)
+        for rfile in self.files_to_download:
+            if self.kill_received:
+                raise Exception('Kill request received, exiting')
+            # Determine where to store file (directory and name)
+            file_dir = local_dir
+            if keep_dirs:
+                file_dir = local_dir + '/' + os.path.dirname(rfile['save_as'])
+            if file_dir[-1] == "/":
+                file_path = file_dir + os.path.basename(rfile['save_as'])
+            else:
+                file_path = file_dir + '/' + os.path.basename(rfile['save_as'])
+
+            # For unit tests only, workflow will take in charge directory
+            # creation before to avoid thread multi access
+            if not os.path.exists(file_dir):
+                os.makedirs(file_dir)
+
+            msg = self.__class__.__name__ + ':Download:Progress:'
+            msg += str(cur_files) + '/' + str(nb_files)
+            msg += ' downloading file ' + rfile['name'] + ' save as ' + rfile['save_as']
+            self.logger.debug(msg)
+            cur_files += 1
+            start_time = datetime.datetime.now()
+            start_time = time.mktime(start_time.timetuple())
+            error = self._download(file_path, rfile)
+            if error:
+                rfile['download_time'] = 0
+                rfile['error'] = True
+                raise Exception(self.__class__.__name__ + ":Download:Error:" + rfile["name"])
+            else:
+                end_time = datetime.datetime.now()
+                end_time = time.mktime(end_time.timetuple())
+                rfile['download_time'] = end_time - start_time
+            # This was moved from FTPDownload
+            if not error and skip_check_uncompress is None:
+                archive_status = Utils.archive_check(file_path)
+                if not archive_status:
+                    self.logger.error('Archive is invalid or corrupted, deleting file and retrying download')
+                    error = True
```

Comment:
This was set in curl_download and returns an error if archive is corrupted.
What will happen here if an error is detected? Not marked as failed and no retry

Reply:
You are right. Our goal was to have the same beahviour for all downloaders.

An easy fix is to raise an exception in `download` (as done when `_download` returns `True`) so the whole process is retried.

We can also retry in `_download` but this is a bit more complex to implement.

Reply:
An error/exception  needs indeed to be raised.
If error was true, file was also deleted (in ftp case at least)

Retry is done at download step in case if error?

Reply:
I think that of `download` fails (i.e. raise an exception), BioMAJ will restart at the download step. @lburlot can you confirm that ?

If I'm correct 25a6c32 should be enough for that.

Note that we can such errors (corrupted in archive) when we will implement retrying of download inside the method (as opposed to BioMAJ restart).